docs/metrics/perplexity.md

```diff
@@ -0,0 +1,142 @@
+perplexity assesses the "confidence" or "surprise" of a language model in predicting the next word in a sequence of words.
+
+Perplexity quantifies how well a language model can predict the next word in a sequence of words. It is calculated
+based on the probability distribution of words generated by the model. A high perplexity indicates that the
+language model is not confident in its text generation — that is, the model is "perplexed" — whereas a low
+perplexity indicates that the language model is confident in its generation.
+
+??? note "Note: Perplexity can only be calculated for autoregressive language models"
+    Autoregressive language models, such as GPT and LLaMA, are language models that work by generating one token at a time, based on a set number of preceding tokens. Simply put, it is a model that outputs a probability distribution of words
+    given some preceding context, selects the word with the highest probability to be the next word in the sentence,
+    then reruns the same process with the generated word as part of the preceding context.
+
+    For example, if the preceding generated text is "Hello, today is a great day to" and our context length is set
+    to 6 preceding words, our model's output distribution might look like the following:
+
+    P(`"walk"` | `"today is a great day to"`) = 46% <br>
+    P(`"run"` | `"today is a great day to"`) = 44% <br>
+    P(`"sleep"` | `"today is a great day to"`) = 9.9% <br>
+    P(`"goodbye"` | `"today is a great day to"`) = 0.1% <br>
+
+    For a deep dive into autoregressive language models like GPT and LLaMA, feel free to refer to [this publication](https://insights.sei.cmu.edu/blog/harnessing-the-power-of-large-language-models-for-economic-and-social-good-foundations/) on the theory behind their inner-workings.
+
+## Implementation Details
+
+Mathematically, perplexity is defined as the exponential of the negative log-likelihood of a sequence (don't worry
+if this doesn't make sense, we'll break it down!).
+
+$$
+\begin{align*}
+\text{Perplexity} &= \exp{\{ -\frac{1}{t} \sum^t_i \log p_\theta (x_i | x_\text{context}) \}}
+\end{align*}
+$$
+
+where $x_i$ represents the token that is generated, and $x_\text{context}$ represents the preceding tokens that the
+current generated token is conditioned on.
+
+In pseudocode, the algorithm works as such:
+
+``` py title="Perplexity Pseudocode"
+total_log_likelihood = 0
+word_count = 0
+
+for example in test_dataset:
+    # The correct next word to be predicted
+    reference_word = example.reference
+    # Predicted output distribution
+    predicted_distribution = model.predict(example.context)
+
+    # Add the log-likelihood of the correct next word predicted by the model.
+    total_log_likelihood += log(predicted_distribution[reference_word])
+    word_count += 1
+
+# Calculate the exponential of the negative log likelihood averaged over the
+#     number of words
```

Comment:
<img width="711" alt="image" src="https://github.com/kolenaIO/kolena/assets/17771952/17f0d93a-7060-4d92-ba5b-ab274175c767">
Is the indentation here intentional?

Reply:
It was intentional to show that the comment continues on the next line, but I've removed the indent since it's still clear without the indent